### PR TITLE
ENH: constructors for sparse arrays 

### LINKED ISSUE #18592: Semantics of sparse array creation functions
https://github.com/scipy/scipy/pull/18538 outlines a procedure for (eventually) replacing existing sparse matrix creation functions with their array equivalents. One of the advantages of the proposed approach is that we get to re-examine the interface of those functions in an array context. This issue is to track how the new interface should look.

- `def identity(n, dtype='d', format=None)`

  The 2D identity matrix is by far the most common case, and this API already follows `np.identity`. It can be kept as-is, with the option of adding an `ndim` keyword later.

- `def eye(m, n=None, k=0, dtype=float, format=None)`

  This function has no natural analog in 1D or ND, so can be kept as-is. Nowadays, we tend to allow `func(shape, ...)` instead of `func(s0, s1, ...)`, so we could allow `np.eye((3, 4))` and eventually deprecate the `n` keyword.

- `hstack(blocks, format=None, dtype=None)`, `vstack`: keep as-is; may need to fail for higher dimensional inputs, unless bmat can be generalized. Does not even need an `_array` equivalent, since the inputs determine the output class.

- `block_diag(mats, format=None, dtype=None)`: keep as-is, output class determined by input.

- `random(m, n, density=0.01, format='coo', dtype=None, random_state=None, data_rvs=None)` : shape should be condensed into a single argument; `random_state` is used to sample sparsity structure; unless `_rvs` is a standard pattern, can consider a more intuitive naming for `data_rvs`? The docstring can also be improved, both in description and example, where e.g. one would expect an example such as `data_rvs=partial(rng.poisson, 0.8)`.

- `rand`: deprecate

- `bmat(blocks, format=None, dtype=None)`: provide `barray` with same semantics.

The following functions are listed under sparse creation utilities, but are operations on sparse objects:

```
   kron - kronecker product of two sparse matrices
   kronsum - kronecker sum of sparse matrices
   tril - Lower triangular portion of a matrix in sparse format
   triu - Upper triangular portion of a matrix in sparse format
```

I.e., can be kept as is, with output arrays as appropriate.

Once we've agreed on all interface decisions above, the implementation should be straightforward.

/cc @rossbar @perimosocordiae

### PR DESCRIPTION
**Pull request by dschult**
I believe this provides the last part of the "Semantics of sparse arrays" issue.
Fixes #18592 

Other partial fixes of this issue have been included in #18538 and #18862 

This PR adds `eye_array` and `random_array` and adapts `kron` and `kronsum` to use sparse arrays when any of the inputs are sparse arrays. This completes the revision of the `_construct.py` module and the resulting functions are:

```python
# New functions recommended along with matrices being discouraged
def diags_array(diagonals, /, *, offsets=0, shape=None, format=None, dtype=None):
def eye_array(m, n=None, *, k=0, dtype=float, format=None):
def random_array(m, n, density=0.01, format='coo', dtype=None, random_state=None, data_random_state=None):
def block(blocks, format=None, dtype=None):

# Existing names that now return sparray or spmatrix dependent on input.
def kron(A, B, format=None):
def kronsum(A, B, format=None):
def hstack(blocks, format=None, dtype=None):
def vstack(blocks, format=None, dtype=None):
def block_diag(mats, format=None, dtype=None):
def tril(A, k=0, format=None):
def triu(A, k=0, format=None):

# Names that should be rewritten to return sparse arrays when sparse matrices are removed
def eye(m, n=None, k=0, dtype=float, format=None):
def random(m, n, density=0.01, format='coo', dtype=None,  random_state=None,  data_rvs=None):
def diags(diagonals, offsets=0, shape=None, format=None, dtype=None):

# Names that should be removed when sparse matrices are removed
def bmat(blocks, format=None, dtype=None):
def rand(m, n, density=0.01, format="coo", dtype=None, random_state=None):
def spdiags(data, diags, m=None, n=None, format=None):
def identity(n, dtype='d', format=None):
```

I have added warnings to doc_strings of discouraged functions recommending sparse arrays. These functions have warnings:
```
bmat
diags
spdiags
identity
eye
random
rand
```

Some changes that might need discussion:
- The new functions have keyword only arguments enforced. 
- For `random_array`, I've changed the keyword ``data_rvs`` to ``data_random_state``. though I would prefer ``data_rng``, I think `data_random_state` matches the library better.
- Should we deprecate some of these? If so, when?
- Are there other parts of the new function API that we should tweak while we're making this change?

### CODE REVIEW

**Comment by miikkuu** [comment]
It seems like the issue here is with the terminology used in the error message. In newer versions of SciPy, the term "sparse matrix" is used instead of "sparse array." The functionality should still be the same.

You can update your test code to expect a "sparse matrix" instead of a "sparse array" and consider it as an acceptable change in terminology.Or make change in your code .

The second error is due to : scipy.sparse.random_array
the error message indicates a NameError for the name 'random,' which suggests that the 'random' function is not defined in the current scope.

To use the random function from SciPy's sparse module, you should import it explicitly. Add the following import statement at the beginning of your code:

from scipy.sparse import random

**Comment by rgommers** [comment]
> The new functions have keyword only arguments enforced.

This is generally desirable in new functionality, so +1 from me.

> * For `random_array`, I've changed the keyword `data_rvs` to `data_random_state`. though I would prefer `data_rng`, I think `data_random_state` matches the library better.

https://github.com/scientific-python/specs/pull/180 is leaning towards `rng` as the future canonical name, and I agree that `data_rng` is a nicer name.

> Should we deprecate some of these? If so, when?

I don't think so, not in this PR or in 1.12.0 at least. For two reasons:
1. Deprecating a function in the same release in which the replacement becomes available forces users to implement version-based if/else usage, so this is usually best avoided,
2. I really would like to see a comprehensive proposal for deprecating sparse matrix classes, as mentioned several times before. Before that gets written, I'd prefer to not see any individual deprecations nor definitive statements that the matrix classes are being or about to be deprecated.

File: scipy/sparse/__init__.py [hunk 1]
```diff
@@ -86,11 +86,13 @@
 .. autosummary::
    :toctree: generated/
 
+   eye_array - Sparse MxN array whose k-th diagonal is all ones
    eye - Sparse MxN matrix whose k-th diagonal is all ones
    identity - Identity matrix in sparse format
    kron - kronecker product of two sparse matrices
    kronsum - kronecker sum of sparse matrices
    diags - Return a sparse matrix from diagonals
+   diags_array - Return a sparse matrix from diagonals
```
Comment (perimosocordiae) [inline scipy/sparse/__init__.py:95]: sparse matrix -> sparse array

File: scipy/sparse/_base.py [hunk 2]
```diff
@@ -353,11 +353,11 @@ def imag(self):
...
     def __repr__(self):
-        _, format_name = _formats[self.format]
+        _, format = _formats[self.format]
         sparse_cls = 'array' if isinstance(self, sparray) else 'matrix'
-        return f"<%dx%d sparse {sparse_cls} of type '%s'\n" \
-               "\twith %d stored elements in %s format>" % \
-               (self.shape + (self.dtype.type, self.nnz, format_name))
+        return (f"<%dx%d sparse {sparse_cls} of type '{self.dtype.type}'\n"
+                f"\twith {self.nnz} stored elements in {format} format>"
+               ) % self.shape
```
Comment (perimosocordiae) [inline scipy/sparse/_base.py:360]: I'm not sure why this change was included in this PR, but we should probably avoid mixing f-string with %-interpolation.

File: scipy/sparse/tests/test_construct.py [hunk 3]
```diff
@@ -302,18 +317,44 @@ def test_kron(self):
...
+        assert_array_equal(result.toarray(), expected)
+        assert isinstance(result, spmatrix)
 
     def test_kron_large(self):
         n = 2**16
-        a = construct.eye(1, n, n-1)
-        b = construct.eye(n, 1, 1-n)
+        # pre-sparse-array setup
+        # a = construct.eye(1, n, n-1)
+        # b = construct.eye(n, 1, 1-n)
```
Comment (perimosocordiae) [inline scipy/sparse/tests/test_construct.py:355]: The commented-out lines should be removed.

File: scipy/sparse/_construct.py [hunk 4]
```diff
@@ -361,57 +451,70 @@ def kron(A, B, format=None):
...
+    >>> sp.sparse.kron(A, [[1, 2], [3, 4]]).toarray()
     array([[ 0,  0,  2,  4],
            [ 0,  0,  6,  8],
            [ 5, 10,  0,  0],
            [15, 20,  0,  0]])
 
     """
-    B = coo_matrix(B)
-
+    # delete next 10 lines and replace _sparse with _array when spmatrix removed
```
Comment (perimosocordiae) [inline scipy/sparse/_construct.py:483]: I'd add the TODO: prefix here to make it more clear that this comment is about future work.

File: scipy/sparse/_construct.py [hunk 5]
```diff
@@ -315,15 +399,21 @@ def eye(m, n=None, k=0, dtype=float, format=None):
...
-    >>> sparse.eye(3).toarray()
+    >>> import scipy as sp
+    >>> sp.sparse.eye(3).toarray()
     array([[ 1.,  0.,  0.],
            [ 0.,  1.,  0.],
            [ 0.,  0.,  1.]])
-    >>> sparse.eye(3, dtype=np.int8)
+    >>> sp.sparse.eye(3, dtype=np.int8)
     <3x3 sparse matrix of type '<class 'numpy.int8'>'
         with 3 stored elements (1 diagonals) in DIAgonal format>
```
Comment (perimosocordiae) [inline scipy/sparse/_construct.py:419]: I'd prefer to avoid duplication of implementation for eye + eye_array. Maybe we can do something similar to `_random()` below?

**Review by perimosocordiae — CHANGES_REQUESTED** [review]

**Review by dschult — COMMENTED** [review]
Thanks for these comments!

I've made the suggested changes. 

I also did some more updates to `sp.sparse.__init__.py`. I'm sure even more is needed (e.g. in the "usage" section). But I hope this makes the part about constructing matrices and arrays better. That process also helped me realize that `load_npz` should probably be given an array version, so I added `load_npz_array`. I believe the function `save_npz` can work as is with either array or matrix. No need to even store what type was saved. We just load the file to the type we want.

After revisiting the `coo-1d` branch and it's handling of the `__repr__` functions I have adopted those format strings from `_base.py` and copied/adapted them for `_dia.py` and `_bsr.py`.

**Comment by dschult** [comment]
I have adapted the format strings for `__repr__` from the `coo-1d` branch. `_base.py` should match that exactly and I copied and adapted for `_dia.py` and `_bsr.py`.

I also rebased on main to resolve conflicts with the `hstack`, etc PR that has been merged.
So this should be ready for review.

New since most recent review is also rewrites of `save_npz` and `load_npz` and addition of `load_npz_array`.

File: scipy/sparse/tests/test_construct.py [hunk 6]
```diff
@@ -225,7 +233,8 @@ def test_diags_empty(self):
         x = construct.diags([])
         assert_equal(x.shape, (0, 0))
 
-    def test_identity(self):
+    @pytest.mark.parametrize("identity", [construct.identity, construct.eye_array])
+    def test_identity(self, identity):
         assert_equal(construct.identity(1).toarray(), [[1]])
```
Comment (perimosocordiae) [inline scipy/sparse/tests/test_construct.py:238]: Should this be using the `identity` argument instead of `construct.identity`?

**Review by perimosocordiae — APPROVED** [review]

File: scipy/sparse/tests/test_construct.py (hunk 6 above)
Comment (dschult) [reply scipy/sparse/tests/test_construct.py:238]: Yes, indeed it should. :)

**Comment by perimosocordiae** [comment]
CI failures seem to be unrelated.

**Comment by dschult** [comment]
This PR now has all green checks. :)
It's also got a +1 at least for the name choices, and owners approval (even though the check for owners approval says it is still needed).   If another set of eyes are needed let me know.
Thanks

File: scipy/sparse/_matrix_io.py [hunk 7]
```diff
@@ -1,15 +1,15 @@
...
 
-__all__ = ['save_npz', 'load_npz']
+__all__ = ['save_npz', 'load_npz', 'load_npz_array']
 
 
 # Make loading safe vs. malicious input
 PICKLE_KWARGS = dict(allow_pickle=False)
 
 
 def save_npz(file, matrix, compressed=True):
```
Comment (stefanv) [inline scipy/sparse/_matrix_io.py:11]: Could we include metadata to indicate which container was stored? It may still make sense to have a flag / example to show how to force a specific format.

```
In [51]: np.savez('/tmp/foo.npz', __format__='array')

In [52]: np.load('/tmp/foo.npz')['__format__']
Out[52]: array('array', dtype='<U5')
```

File: scipy/sparse/_construct.py [hunk 8]
```diff
@@ -1048,45 +1173,147 @@ def random(m, n, density=0.01, format='coo', dtype=None,
...
         elif np.issubdtype(dtype, np.complexfloating):
-            def data_rvs(n):
+            def data_rng(n):
                 return (random_state.uniform(size=n) +
                         random_state.uniform(size=n) * 1j)
         else:
-            data_rvs = partial(random_state.uniform, 0., 1.)
+            data_rng = partial(random_state.uniform, 0., 1.)
 
     ind = random_state.choice(mn, size=k, replace=False)
```
[the remark this answers is not in the document]
Comment (dschult) [reply scipy/sparse/_construct.py:1202]: This raises exceptions or gives duplicate entries (because `replace=False` only applies across 1 axis instead of all axes). I think a fast unravel-er like `np.unravel_index` could be worthwhile:
```python
    indices = np.unravel_index(random_state.choice(np.prod(shape), size=k, replace=False), shape=shape)
```

[the remark this answers is not in the document]
Comment (stefanv) [reply scipy/sparse/_construct.py:1202]: That's kind of weird, right, one would expect it to be totally feasible to sample with replacement across N dimensions. Perhaps a feature we should add to NumPy!

[the remark this answers is not in the document]
Comment (stefanv) [reply scipy/sparse/_construct.py:1202]: https://mail.python.org/archives/list/numpy-discussion@python.org/thread/O7CBSJU5I2PXMOBZC7MQ7FKFSO33E3FS/

With thanks to @rkern.

If we're happy with the uint64 upper limited, going with the `choice` solution works. At least we have options, should this become an issue later.

[the remark this answers is not in the document]
Comment (dschult) [reply scipy/sparse/_construct.py:1202]: The current code uses dtype `np.intc`. I don't know if people are running into that limit or not...
The choice code is something like this:
```python
     mn = np.prod(shape)
     size = int(round(density * mn))
     raveled_ind = rng.choice(np.prod(shape), size=size, replace=False).astype(dtype)
     indices = np.unravel_index(raveled_ind, shape=shape)
```
The integers method is slower maybe somewhat because of repacking after being generated.
```python
     mn = np.prod(shape)
     size = int(round(density * mn))
     seen = set()
     while len(seen) < size:
         dsize = size - len(seen)
         seen.update(map(tuple, rng.integers(shape, size=(dsize, ndim), dtype=dtype)))
     indices = np.array(list(seen), dtype=dtype).T
```
Nice versions for ints larger than 2**64 but that's probably not needed I'm guessing.
:)

[the remark this answers is not in the document]
Comment (stefanv) [reply scipy/sparse/_construct.py:1202]: No, but probably worth ensuring that prod(shape) fits in intc. uint always makes more sense, int64 may make more sense sometimes?

[the remark this answers is not in the document]
Comment (dschult) [reply scipy/sparse/_construct.py:1202]: Right -- the max value has to fit...

The `choice` version is about 30 times faster in my tests for `size = 200-20000`
(when `np.prod(shape)` is small enough that it can work). 
So I propose a hybrid method depending on whether `math.prod(shape) < np.iinfo('uint64').max`.

Also, are we supposed to use `scipy.utils.rng_integers` instead of `rng.integers` so it works with a `random_state` input?

[the remark this answers is not in the document]
Comment (rkern) [reply scipy/sparse/_construct.py:1202]: FYI, `choice()` is restricted to `mn` being an `int64`.

File: scipy/sparse/_construct.py [hunk 9]
```diff
@@ -953,10 +1064,12 @@ def block_diag(mats, format=None, dtype=None):
                       dtype=dtype).asformat(format)
 
 
-def random(m, n, density=0.01, format='coo', dtype=None,
-           random_state=None, data_rvs=None):
-    """Generate a sparse matrix of the given shape and density with randomly
-    distributed values.
+def random_array(m, n=None, *, density=0.01, format='coo', dtype=None,
```
[the remark this answers is not in the document]
Comment (dschult) [reply scipy/sparse/_construct.py]: If we are changing the args of `random_array` (compared to `random`) should we rename `random_state` to `rng`? or is that a different scale of api shift?

File: scipy/sparse/_construct.py [hunk 10]
```diff
@@ -988,55 +1102,66 @@ def random(m, n, density=0.01, format='coo', dtype=None,
...
-        This random state will be used for sampling the sparsity structure, but
-        not necessarily for sampling the values of the structurally nonzero
-        entries of the matrix.
-    data_rvs : callable, optional
+        This random state will be used for sampling `indices` (the sparsity
+        structure), and possibly for the data values themselves, but if you
+        prefer another random state for data (e.g. normal distribution)
+        use ``data_rng`` and see the examples below.
+
+    data_rng : callable, optional
```
[the remark this answers is not in the document]
Comment (dschult) [reply scipy/sparse/_construct.py]: The code used to call the `data_rvs` function with a position argument holding the size. That makes it so we have to make the distribution `frozen`. It also means we can't use `partial` from the numpy random distributions (the position argument gives a duplicate value for the first argument).  So I've changed the code to call it as a keyword arg:  `data_sampler(size=size)`. That allows sp.stats distributions even when not frozen, and `partial` versions of np.random distributions.

Open to renaming `data_sampler`, but it hopefully makes more clear that we don't have two RNGs, we have one RNG and a function that creates samples from a distribution. The default data_samplers have distributions based on dtype and use the RNG to produce them.

File: scipy/sparse/_construct.py [hunk 11]
```diff
@@ -978,8 +1091,9 @@ def random(m, n, density=0.01, format='coo', dtype=None,
         sparse matrix format.
     dtype : dtype, optional
         type of the returned matrix values.
-    random_state : {None, int, `numpy.random.Generator`,
-                    `numpy.random.RandomState`}, optional
+    random_state : {None, int, `Generator`, `RandomState`}, optional
```
[the remark this answers is not in the document]
Comment (dschult) [reply scipy/sparse/_construct.py:1092]: Should this argument be renamed `rng=`?  I left it as random_state to match the rest of the library. But this **is** a new function so we could break from using `check_random_state` and move to the rng interface.

Also, to go another step in the process, we could make the default value be `np.random.default_rng()`.  Is this too much change? We don't get to introduce new functions for old functionality very often.

File: scipy/sparse/_matrix_io.py (hunk 7 above)
Comment (dschult) [reply scipy/sparse/_matrix_io.py:11]: I added a key-value pair to the npz file with key `__is_array` and value `array` or `matrix`. This gets concatenated with `self.format` to create the class constructor `sp.sparse.<format>_<array or matrix>`.
Other choices are certainly possible. I also added an example which takes the output based on `__is_array` and changes it by feeding it into a `csr_array` constructor.

File: scipy/sparse/_construct.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dschult) [reply scipy/sparse/_construct.py]: I've implemented a hybrid approach dependent on the size of `math.prod(shape)`. I used the `int64` maximum in the code rather than `uint64` because that matches scipy/sparse's choice for `idx_dtype` throughout (either int32 or int64). I'm not sure of the history there.

File: scipy/sparse/_construct.py [hunk 12]
```diff
@@ -283,54 +306,74 @@ def identity(n, dtype='d', format=None):
...
+    Return a sparse array with ones on diagonal.
+    Specifically a sparse array (m x n) where the kth diagonal
     is all ones and everything else is zeros.
 
     Parameters
     ----------
-    m : int
-        Number of rows in the matrix.
+    m : int or tuple of ints
+        Number of rows or shape requested.
```
Comment (stefanv) [inline scipy/sparse/_construct.py:335]: Double checking if "or shape requested" is correct here.

File: scipy/sparse/_construct.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (stefanv) [reply scipy/sparse/_construct.py:1095]: Since it's a new function, I hope we can use the most likely future-compatible API. @perimosocordiae?

File: scipy/sparse/_construct.py [hunk 13]
```diff
@@ -953,6 +1065,160 @@ def block_diag(mats, format=None, dtype=None):
...
+
+    data_sampler : callable, optional (default depends on dtype)
+        Sampler of random data values with keyword arg `size`.
+        This function should take a single keyword argument `size` specifying
+        the length of its returned ndarray. It is used to generate the nonzero
+        values in the matrix after the locations of those values are chosen.
+        By default, uniform [0, 1) random values are used unless `dtype` is
+        an integer (default uniform integers from that dtype) or
+        complex (default uniform over the unit square in the complex plane).
+        Also by default the `random_state` rng is used e.g.
```
Comment (stefanv) [inline scipy/sparse/_construct.py:1118]: This sentence really begs changing to `rng.uniform`.

File: scipy/sparse/_construct.py [hunk 14]
```diff
@@ -953,6 +1065,160 @@ def block_diag(mats, format=None, dtype=None):
...
+        ind = np.unravel_index(raveled_ind, shape=shape)
+    else:
+        # for ravel indices bigger than dtype max, use sets to remove duplicates
+        ndim = len(shape)
+        seen = set()
+        while len(seen) < size:
+            dsize = size - len(seen)
+            seen.update(map(tuple, rng_integers(rng, shape, size=(dsize, ndim))))
+        ind = tuple(np.array(list(seen)).T)
+        print("size(ind): ",tuple(a.shape for a in ind))
```
Comment (stefanv) [inline scipy/sparse/_construct.py:1215]: debug statement

File: scipy/sparse/_matrix_io.py [hunk 15]
```diff
@@ -67,14 +68,16 @@ def save_npz(file, matrix, compressed=True):
         shape=matrix.shape,
         data=matrix.data
     )
+    if isinstance(matrix, sp.sparse.sparray):
+        arrays_dict.update(__is_array="array")
```
Comment (stefanv) [inline scipy/sparse/_matrix_io.py:72]: It doesn't really matter, but does `__is_array=True` not work?

File: scipy/sparse/_matrix_io.py [hunk 16]
```diff
@@ -94,58 +97,68 @@ def load_npz(file):
...
+    >>> sp.sparse.save_npz('/tmp/sparse_matrix.npz', sparse_matrix)
+    >>> tmp = sp.sparse.load_npz('/tmp/sparse_matrix.npz')
+    >>> sparse_array = sp.sparse.csr_array(tmp)
+    """
     with np.load(file, **PICKLE_KWARGS) as loaded:
         try:
-            matrix_format = loaded['format']
+            sparse_format = loaded['format']
         except KeyError as e:
             raise ValueError(f'The file {file} does not contain a sparse matrix.') from e
```
Comment (stefanv) [inline scipy/sparse/_matrix_io.py:137]:
```suggestion
            raise ValueError(f'The file {file} does not contain a sparse array or matrix.') from e
```

Comment (stefanv) [inline scipy/sparse/_matrix_io.py:137]: Nothing wrong with this, but feels a bit convoluted. I know it's not yours :) How about:

```
sparse_format = loaded.get('format')
if sparse_format is None:
   ...
```

File: scipy/sparse/_matrix_io.py [hunk 17]
```diff
@@ -94,58 +97,68 @@ def load_npz(file):
...
+        sparse_format = sparse_format.item()
 
-        if not isinstance(matrix_format, str):
+        if not isinstance(sparse_format, str):
             # Play safe with Python 2 vs 3 backward compatibility;
             # files saved with SciPy < 1.0.0 may contain unicode or bytes.
-            matrix_format = matrix_format.decode('ascii')
+            sparse_format = sparse_format.decode('ascii')
+
+        if '__is_array' in loaded and loaded['__is_array'].item() == 'array':
```
Comment (stefanv) [inline scipy/sparse/_matrix_io.py:146]:
```
if loaded.get('__is_array'):
```

**Review by stefanv — COMMENTED** [review]
Looks pretty good to me!

@perimosocordiae do you want to cast another eye?

File: scipy/sparse/_construct.py (hunk 13 above)
Comment (dschult) [reply scipy/sparse/_construct.py:1118]: I tried changing it to:
 For these, the `random_state` rng is used e.g. `rng.uniform(size=size)`.

If we rename `random_state` to `rng` then it becomes:
 For the default cases, `rng` is used e.g. `rng.uniform(size=size)`

File: scipy/sparse/_matrix_io.py (hunk 15 above)
Comment (dschult) [reply scipy/sparse/_matrix_io.py:72]: I changed the key to `_is_array` and the value to `True`.

File: scipy/sparse/_construct.py [hunk 18]
```diff
@@ -953,6 +1065,158 @@ def block_diag(mats, format=None, dtype=None):
...
+            random_state=None, data_sampler=None):
+    if density < 0 or density > 1:
+        raise ValueError("density expected to be 0 <= density <= 1")
+
+    tot_prod = math.prod(shape)  # use `math` for when prod is >= 2**64
+
+    # Number of non zero values
+    size = int(round(density * tot_prod))
+
+    rng = check_random_state(random_state)
```
[the remark this answers is not in the document]
Comment (stefanv) [reply scipy/sparse/_construct.py:1185]: else

File: scipy/sparse/_construct.py [hunk 19]
```diff
@@ -953,6 +1065,161 @@ def block_diag(mats, format=None, dtype=None):
...
+    if density < 0 or density > 1:
+        raise ValueError("density expected to be 0 <= density <= 1")
+
+    tot_prod = math.prod(shape)  # use `math` for when prod is >= 2**64
+
+    # Number of non zero values
+    size = int(round(density * tot_prod))
+
+    if random_state is None:
+        random_state = np.random.random_rng()
```
Comment (perimosocordiae) [inline scipy/sparse/_construct.py:1186]: this is supposed to be `np.random.default_rng()`

**Comment by perimosocordiae** [comment]
Looks like the test failures are real:

 - scipy/sparse/linalg/_eigen/arpack/tests/test_arpack.py::test_real_eigs_real_k_subset
 - scipy/sparse/linalg/_isolve/tests/test_gcrotmk.py::TestGCROTMK::test_arnoldi

**Comment by perimosocordiae** [comment]
Ok, I pushed a fix that works for me locally. If the CI agrees, I think this is ready to merge now.

**Comment by dschult** [comment]
That'll teach me to try making PRs on a browser during a meeting. :)

**Comment by perimosocordiae** [comment]
CI failure is unrelated (import error related to scipy.cluster), so I'll merge now.

**Comment by stefanv** [comment]
:tada: Thanks Dan and CJ!
